service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java

```diff
@@ -234,11 +258,116 @@ private static void processRequestHeaders(HttpAcceptSession acceptSession, HttpC
 
         // Add Via: 1.1 kaazing header
         connectSession.addWriteHeader(HEADER_VIA, VIA_HEADER_VALUE);
+
+        // Add forwarded headers
+        setupForwardedHeaders(acceptSession, connectSession, forwardedProperty);
+
     }
-    
+
+    /**
+     * Compose the Forwarded header and the X-Forwarded headers and add them to the connect session write headers.
+     * 
+     * @param acceptSession
+     * @param connectSession
+     * @param forwardedProperty the value of the 'use-forwarded' property used for http.proxy type in gateway-config:
+     * inject ( add the corresponding data to the Forwarded/X-Forwarded headers ), ignore ( this proxy is anonymous, no
+     * forwarded header data is added ), or exclude ( delete any existing Forwarded/X-Forwarded headers received, and do
+     * not add any new data )
+     */
+    private static void setupForwardedHeaders(HttpAcceptSession acceptSession,
+                                              HttpConnectSession connectSession,
+                                              String forwardedProperty) {
+        if (FORWARDED_EXCLUDE.equalsIgnoreCase(forwardedProperty)) {
+            excludeForwardedHeaders(connectSession);
+            return;
+        }
+        StringBuilder forwardedSb = new StringBuilder();
+        String forwarded = acceptSession.getReadHeader(HEADER_FORWARDED);
+        String remoteIpAddress = getResourceIpAddress(acceptSession, FORWARDED_FOR);
+        if (remoteIpAddress != null) {
+            // 'Forwarded: for=' is added to the connectSession in HttpSubjecSecurityFilter if the 'Forwarded' header in
+            // the request is empty, so we remove it for now, and inject it along with all the other parameters (by,
+            // proto, host, port)
+            String forAddress = format("%s=%s", FORWARDED_FOR, remoteIpAddress);
+            if (forwarded.equalsIgnoreCase(forAddress)) {
+                connectSession.clearWriteHeaders(HEADER_FORWARDED);
+            }
+            if (FORWARDED_INJECT.equalsIgnoreCase(forwardedProperty)) {
+                connectSession.addWriteHeader(HEADER_X_FORWARDED_FOR, remoteIpAddress);
+                forwardedSb.append(forAddress + ";");
+            }
+        }
+
+        if (FORWARDED_INJECT.equalsIgnoreCase(forwardedProperty)) {
+
+            String serverIpAddress = getResourceIpAddress(acceptSession, FORWARDED_BY);
+            if (serverIpAddress != null) {
+                connectSession.addWriteHeader(HEADER_X_FORWARDED_SERVER, serverIpAddress);
+                forwardedSb.append(format("%s=%s;", FORWARDED_BY, serverIpAddress));
+            }
+
+            URI externalURI = acceptSession.getLocalAddress().getExternalURI();
+            String protocol = externalURI.getScheme();
+            connectSession.addWriteHeader(HEADER_X_FORWARDED_PROTO, protocol);
+            forwardedSb.append(format("%s=%s;", FORWARDED_PROTO, protocol));
```

Comment:
Just because, scheme is https doesn't mean the protocol is https. The underlying stack may be configured and replaced differently (may be you should use acceptSession.isSecure() for http or https)